### DISCUSSION TITLE
Any plan for AI assistant to be on the roadmap?

### DISCUSSION THREAD

**Discussion by kingfungdesign** [outline/outline#4972] [General]
Having tried ChatGPT and Notion AI, I can confidently say that AI will be a tool that everyone will use to assist with writing and content generation in the very near future. It can easily double productivity, and those who do not use it will be left far behind.

Now that OpenAI has launched an official API for GPT, will implementing AI assistance be part of Outline's roadmap? I would love to support Outline even if it costs extra and have AI-assisted writing in one place, rather than having to use Notion or ChatGPT for AI-assisted writing separately.

**Comment by ChuckJonas** [comment]
If this gets added, it would be ideal to be able to provide your own Open-AI key.

**Reply by tommoor** [reply]
I already built the proof of concept, its pretty great.

**Reply by kingfungdesign** [reply]
Woah, that's great news!

**Reply by breart** [reply]
Looking forward to it!

**Reply by justinr1234** [reply]
> I already built the proof of concept, its pretty great.

Do you have a branch?

**Reply by therajumandapati** [reply]
@tommoor Was the POC you implemented for Generative use-cases?

**Comment by sergi0g** [comment]
Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.

**Reply by Limezy** [reply]
Thanks, you wrote the comment I was about to make 😂

**Reply by tommoor** [reply]
> Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.

Care to say why?

I should clarify, right now we're not looking at generative AI but using vectors to implement true semantic search and summarization, and yes it would be optional as it requires a different version of Redis.

**Reply by kingfungdesign** [reply]
> > Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.
> 
> Care to say why?
> 
> I should clarify, right now we're not looking at generative AI but using vectors to implement true semantic search and summarization, and yes it would be optional as it requires a different version of Redis.

Please also consider functions such as "improve writing" and "fix grammar and spelling" to be of higher priority. They are really helpful for proofreading and for helping to rephrase paragraphs into a more concise and easy-to-read format, especially for non-native speaker. It just saves me so much time to just write things out and let the AI do the tidying for me.

"Translate..." and "Suggest me some ideas about..." are also ones that I find very useful.

> Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.

I'm also curious. I do think there should be an option to opt-out, and if possible, make it a per-member toggle with permission settings. If Outline ever considers making it a paid feature, it should be billed per-member-toggled too. Notion's approach of forcing payment for all workspace members, even if only one person uses the feature, feels really dumb.
I use Notion everyday for work and I've never find the Ai feature to be intruding in any way nor have I ever accidently trigger it.

**Reply by sergi0g** [reply]
Good to know, thanks. I don't need an AI when writing and hate the "Press space for AI" thing Notion implemented. If I ever need an AI, opening the ChatGPT website is just as easy. I want to use Outline for keep track of tasks, so the API feature is really helpful and for writing notes occasionally.

**Reply by Limezy** [reply]
> > Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.
> 
> Care to say why?
> 
> I should clarify, right now we're not looking at generative AI but using vectors to implement true semantic search and summarization, and yes it would be optional as it requires a different version of Redis.

With the clarification it looks desirable indeed !

**Reply by justinr1234** [reply]
> > Please don't implement this, or make it optional, because Notion's addition of AI without any opt-out is what made me switch to outline.
> 
> Care to say why?
> 
> I should clarify, right now we're not looking at generative AI but using vectors to implement true semantic search and summarization, and yes it would be optional as it requires a different version of Redis.

I’m confused. Are you saying it won’t work like Notion?

Being able to generate stuff is a huge feature of Notion AI.

**Comment by michael-gratton** [comment]
AI mansplaining is literally the last thing I want when writing detailed technical documentation in Outline.

**Comment by mweth** [comment]
Adding my strong support for this.  If we don't get on the bus, we will get left behind.

**Comment by Mekacher-Anis** [comment]
Just to add my two cents in here, for the people that are strongly against this, remember it's not impossible to make a feature opt-out/in
Some of the featuers I can think of for AI in outline:

- Rephrasing / formatting / proof-reading
- Better semantic search using sentence embeddings
- Summarizing stuff
- Question answering: answers your question while referencing the relevant information in the knowledge-base (not just display links)

**Comment by jaredcwhite** [comment]
Absolutely would prefer opt-in for any generative AI features (_at the very least_ opt-out so then it's removed from the UI). I'm actively reviewing all the apps I use and discarding ones which are throwing LLM frontends up in my face. (The reason I'm trying out Outline right now, in fact! ☺️)

**Reply by sergi0g** [reply]
The only AI app I've got is a gpt frontend 😂

**Comment by tommoor** [comment]
We're taking a very pragmatic view at LLM's, they will only be used where real value can be added and functionality will always be able to be toggled off. For example full semantic search is available on cloud-hosted Outline, now but you existing customers must explicitly enable it under **Settings -> Features** .

**Comment by mweth** [comment]
Here are some things that are valuable:

**Page Level Summation**: Offering a concise summary of the page content can help users quickly understand the page's contents without needing to read through everything. Or create an index of key terms and topics covered in a document, with links to the relevant sections for easy navigation.

**Automated Question Answering**: Implement a feature where users can ask questions directly related to the page content, and the LLM provides precise answers based on the information available on the page.

**Contextual Recommendations**: Suggest related pages or documents based on the content of the current page. This can aid in research and information discovery within the platform.

**Content Categorization and Tagging**: Automatically categorize and tag pages based on their content, making it easier for users to find related information across the platform.

**Version Summary**: Provide a summary of changes for each version of a document, helping users quickly understand what was updated without comparing versions manually.

**Drafting Assistance**: Offer suggestions for improving the content, such as rephrasing for clarity, grammar checks, or style consistency.

**Collaboration Facilitation**: Integrate features that assist in collaborative document editing, such as suggesting contributors for specific topics based on their past contributions or expertise.

**Comment by kindrebelG** [comment]
In the mean time, in case folks are interested in _some_ kind of ChatGPT -> Outline integration, it's fairly easy to rig it up to a custom GPT. I have one setup and use it to spin up docs rapidly, edit, and do translations as well. I only set it up for a portion of the Outline API but I've found it pretty handy. Attached is a hacked up version of the OpenAPI spec that you can drop into the GPT creator interface. Could likely be expanded some-- tho I hit limits when trying the full API. Due to GH limitations, it's an md, but the yml is within 😅 
[outlineOpenAI.md](https://github.com/outline/outline/files/14467740/outlineOpenAI.md)

**Reply by tommoor** [reply]
Nice, are you self hosting or using the cloud instance?

**Reply by kindrebelG** [reply]
> Nice, are you self hosting or using the cloud instance?

The cloud instance atm

**Reply by klinshy** [reply]
for a noob like me, how can I get started using this?
my thanks!

**Comment by Peppershade** [comment]
The current AI search is sort of useful, but not really what I want to see. I'd upgrade to a business plan if I was able to run an AI writing assistant in Outline. Currently, it feels like Outline lacks good AI functionality. Preferably, it would work with custom OpenAI API's, like the one from Ollama. Is something like that on the roadmap?

Edit: To be more specific, I'll add examples of how I'd see it work
- AI chatbox like in VSCode copilot (in the right menu of a page while editing) where you can ask questions about the content, or ask the AI to add something to the page. It would then suggest it inline where you can accept it. This could also be done with a slash command.
- Rewrite highlighted text with a button when editing a page

**Comment by mkay1375** [comment]
[Chainlit](https://chainlit.io/) provides a [Copilot platform](https://docs.chainlit.io/deploy/copilot), which we’ve used to integrate our AI app into Outline:

![image_2025-05-10_130052165(1)](https://github.com/user-attachments/assets/e1ef68dd-60ce-498c-921d-55ed19951771)

We implemented it by patching the following files:
- `server/static/index.html`: Added JS scripts to embed the Copilot widget.
- `server/services/web.js`: Configured the CSP to allow our Copilot server URL.
- `server/routes/app.js`: Generated a JWT using the Chainlit secret and injected it into the `index.html` file to authenticate the Copilot.

Although it works, it’s not a clean solution.
I’m willing to create a PR and implement the integration in a cleaner way. I’ve taken a quick look at the project structure, and the `plugins` folder seems like a good starting point.

Any guidance or suggestions would be appreciated.
